qrand/platforms/qsharp/job.py

```diff
@@ -31,34 +35,129 @@
+        backend: QsharpBackend,
+        num_measurements: Optional[int] = None,
+    ) -> None:
+        self.backend = backend
+        self.circuit = circuit
+        self.num_measurements = num_measurements  # type: ignore
 
     ############################### PUBLIC API ###############################
     @property
     def backend(self) -> QsharpBackend:
-        raise NotImplementedError(self.ERROR_MSG)
+        return self._backend
 
     @backend.setter
     def backend(self, backend: QsharpBackend) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._backend: QsharpBackend = backend
 
     @property
     def circuit(self) -> QsharpCircuit:
-        raise NotImplementedError(self.ERROR_MSG)
+        return self._circuit
 
     @circuit.setter
     def circuit(self, circuit: QsharpCircuit) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        if self.backend.max_qubits < circuit.num_qubits:
+            raise RuntimeError(
+                f"Failed to assign QsharpCircuit for QsharpJob. Number of \
+                qubits in QsharpCircuit unsupported by this job's Backend: \
+                {self.backend.max_qubits}<{circuit.num_qubits}."
+            )
+        self._circuit: QsharpCircuit = circuit
 
     @property
     def num_measurements(self) -> int:
-        raise NotImplementedError(self.ERROR_MSG)
+        return self._shots * self._experiments
 
     @num_measurements.setter
     def num_measurements(self, num_measurements: Optional[int]) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        num_measurements = (
+            num_measurements
+            if num_measurements
+            and type(num_measurements) is int
+            and 0 < num_measurements
+            else self.backend.max_measurements
+        )
+        if self.backend.max_measurements < num_measurements:
+            warn(
+                f"Number of measurements unsupported by the job's Backend: \
+                {self.backend.max_measurements}<{num_measurements}. \
+                Using max_measurements instead.",
+                UserWarning,
+            )
+            num_measurements = self.backend.max_measurements
+        self._shots, self._experiments = compute_bounded_factorization(
+            num_measurements,
+            self.backend.max_shots,
+            self.backend.max_experiments,
+        )
+
+    def _generate_code(self) -> QSharpCallable:
+        qsharp_code = """
+        open Microsoft.Quantum.Intrinsic;
+        open Microsoft.Quantum.Measurement;
+        open Microsoft.Quantum.Arrays;
+
+        operation Program():String[]{{
+
+        use q = Qubit[{num_qubits}];
+        mutable value = ConstantArray({num_qubits},"0");
+
+        {gates}
+
+        ResetAll(q);
+        return value;
+        }}
+        """
+        return compile(
+            qsharp_code.format(
+                num_qubits=self.circuit.num_qubits, gates=self.circuit.gates
+            )
+        )
```

Comment:
The qsharp code defined as a Python string in the function has only one operation, `operation Program():String[]`, which compiles and returns only one callable, `Program()` and hence the return type is just `QSharpCallable`.

Reply:
Yes, in this case all you are getting back is one `QSharpCallable`, but still the complete signature of the function is not that. Since you are not converting types (i.e. checking if you receive a list and "converting" it to a single callable) mypy should be sending some error: isn't it?

Reply:
mypy didn't show any error, that's why I was able to commit in the first place.

Reply:
I am wondering how is that the case though, any explanation?

Reply:
@Avhijit-codeboy sent me the source code for `compile` through discord: https://github.com/microsoft/iqsharp/blob/main/src/Python/qsharp-core/qsharp/__init__.py